src/primitives/poseidon.rs

```diff
@@ -267,84 +291,64 @@ pub trait Domain<F: FieldExt, const T: usize, const RATE: usize>: Copy + fmt::De
 #[derive(Clone, Copy, Debug)]
 pub struct ConstantLength<const L: usize>;
 
-impl<F: FieldExt, const T: usize, const RATE: usize, const L: usize> Domain<F, T, RATE>
-    for ConstantLength<L>
-{
-    fn initial_capacity_element(&self) -> F {
+impl<F: FieldExt, const RATE: usize, const L: usize> Domain<F, RATE> for ConstantLength<L> {
+    type Padding = iter::Take<iter::Repeat<F>>;
+
+    fn name() -> String {
+        format!("ConstantLength<{}>", L)
+    }
+
+    fn initial_capacity_element() -> F {
         // Capacity value is $length \cdot 2^64 + (o-1)$ where o is the output length.
         // We hard-code an output length of 1.
         F::from_u128((L as u128) << 64)
     }
 
-    fn padding(&self) -> SpongeState<F, RATE> {
-        // For constant-input-length hashing, padding consists of the field elements being
-        // zero.
-        let mut padding = [None; RATE];
-        for word in padding.iter_mut().skip(L) {
-            *word = Some(F::zero());
-        }
-        padding
-    }
-
-    fn pad_and_add(&self) -> Box<dyn Fn(&mut State<F, T>, &SpongeState<F, RATE>)> {
-        Box::new(|state, input| {
-            // `Iterator::zip` short-circuits when one iterator completes, so this will only
-            // mutate the rate portion of the state.
-            for (word, value) in state.iter_mut().zip(input.iter()) {
-                // For constant-input-length hashing, padding consists of the field
-                // elements being zero, so we don't add anything to the state.
-                if let Some(value) = value {
-                    *word += value;
-                }
-            }
-        })
+    fn padding(input_len: usize) -> Self::Padding {
+        assert_eq!(input_len, L);
+        // For constant-input-length hashing, we pad the input with zeroes to a multiple
+        // of RATE. On its own this is not sponge-compliant padding, but the Poseidon
+        // authors instead encode the constant length into the capacity element, ensuring
```

Comment:
```suggestion
        // of RATE. On its own this would not be sponge-compliant padding, but the
        // Poseidon authors encode the constant length into the capacity element, ensuring
```